services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/oauth/OIDCTokenHelperImpl.java

```diff
@@ -144,52 +121,45 @@ public String createOIDCaccessToken(
 		return createSignedJWT(claims);
 	}
 	
-	public static RSAPublicKey getRSAPublicKeyForJsonWebKeyRSA(JsonWebKeyRSA jwkRsa) {
-		BigInteger modulus = new BigInteger(jwkRsa.getN());
-		BigInteger publicExponent = new BigInteger(jwkRsa.getE());
-		RSAPublicKeySpec keySpec = new RSAPublicKeySpec(modulus, publicExponent);
-		try {
-			KeyFactory kf = KeyFactory.getInstance("RSA");
-			return (RSAPublicKey)kf.generatePublic(keySpec);
-		} catch (NoSuchAlgorithmException | InvalidKeySpecException e) {
-			throw new RuntimeException(e);
-		} 
-	}
-	
 	@Override
-	public Jwt<JwsHeader,Claims> validateJWTSignature(String token) {
+	public Jwt<JwsHeader,Claims> parseJWT(String token) {
 		// This is a little awkward:  We first have to parse the token to
 		// find the key Id, then, once we map the key Id to the signing key,
 		// we parse again, setting the matching public key for verification
 		String[] pieces = token.split("\\.");
-		if (pieces.length!=3) throw new IllegalArgumentException("Expected three pieces but found "+pieces.length);
+		if (pieces.length!=3) throw new IllegalArgumentException("Expected three sections of the token but found "+pieces.length);
 		String unsignedToken = pieces[0]+"."+pieces[1]+".";
 		JsonWebKey matchingKey=null;
 		{
 			Jwt<Header,Claims> unsignedJwt = Jwts.parser().parseClaimsJwt(unsignedToken);
+			String keyId = (String)unsignedJwt.getHeader().get(JwsHeader.KEY_ID);
 			for (JsonWebKey jwk : jsonWebKeySet.getKeys()) {
-				if (jwk.getKid().equals(unsignedJwt.getHeader().get("kid"))) {
+				if (jwk.getKid().equals(keyId)) {
 					matchingKey = jwk;
```

Comment:
Missing break? An alternative is to replace the for loop with a stream filter expression, e.g:
jsonWebKeySet.getKeys().stream().filter(key -> key.getKid().equals(keyId)).findFirst(). Note that it returns an Optional, which can be used later on with ifPresent().

Reply:
I added in the 'break'